controller/system_backup_controller.go

```diff
@@ -0,0 +1,1068 @@
+	cmd.Dir = filepath.Dir(tempDir)
+	err = cmd.Run()
+	if err != nil {
+		errMessage = fmt.Sprintf("%v: %v", SystemBackupErrArchieve, err)
+		return
+	}
+
+	_, err = os.Stat(archievePath)
+	if err != nil {
+		errMessage = fmt.Sprintf("%v: %v", SystemBackupErrOSStat, err)
+		return
+	}
+}
+
+func (c *SystemBackupController) generateSystemBackup(systemBackupMeta *SystemBackupMeta, tempDir string) (err error) {
+	metaFile := filepath.Join(tempDir, "metadata.yaml")
+	err = encodeToYAMLFile(systemBackupMeta, metaFile)
+	if err != nil {
+		return err
+	}
+
+	return c.generateSystemBackupYAMLs(filepath.Join(tempDir, "yamls"))
+}
+
+func (c *SystemBackupController) generateSystemBackupYAMLs(yamlsDir string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, SystemBackupErrGenerateYAML)
+	}()
+
+	schemeGenerateFns := map[string]func(string) error{
+		filepath.Join(yamlsDir, "apiextensions"): c.generateSystemBackupYAMLsForAPIExtensions,
+		filepath.Join(yamlsDir, "kubernetes"):    c.generateSystemBackupYAMLsForKubernetes,
+		filepath.Join(yamlsDir, "longhorn"):      c.generateSystemBackupYAMLsForLonghorn,
+	}
+	for scheme, fn := range schemeGenerateFns {
+		err = fn(scheme)
+		if err != nil {
+			return
+		}
+	}
+	return
+}
+
+func (c *SystemBackupController) generateSystemBackupYAMLsForLonghorn(dir string) (err error) {
+	scheme := runtime.NewScheme()
+	err = longhorn.AddToScheme(scheme)
+	if err != nil {
+		return errors.Wrap(err, "failed to add Longhorn to scheme")
+	}
+
+	resourceGetFns := map[string]func() (runtime.Object, error){
+		"setting":       c.ds.GetAllLonghornSettings,
+		"engineimages":  c.ds.GetAllLonghornEngineImages,
+		"volumes":       c.ds.GetAllLonghornVolumes,
+		"recurringjobs": c.ds.GetAllLonghornRecurringJobs,
+	}
+
+	for name, fn := range resourceGetFns {
+		err = getObjectsAndPrintToYAML(dir, name, fn, scheme)
+		if err != nil {
+			return
+		}
+	}
+
+	return nil
+}
+
+func (c *SystemBackupController) generateSystemBackupYAMLsForKubernetes(dir string) (err error) {
+	scheme := kubernetesscheme.Scheme
+
+	err = c.generateSystemBackupYAMLsForServiceAccount(dir, "serviceaccounts", "clusterroles", "clusterrolebindings", scheme)
+	if err != nil {
+		return
+	}
+
+	err = c.generateSystemBackupYAMLsForPodSecurityPolicy(dir, "roles", "rolebindings", "podsecuritypolicies", scheme)
+	if err != nil {
+		return
+	}
+
+	err = getObjectsAndPrintToYAML(dir, "daemonsets", c.ds.GetAllDaemonSetsList, scheme)
+	if err != nil {
+		return
+	}
+
+	err = getObjectsAndPrintToYAML(dir, "deployments", c.ds.GetAllDeploymentsList, scheme)
+	if err != nil {
+		return
+	}
+
+	err = getObjectsAndPrintToYAML(dir, "configmaps", c.ds.GetAllConfigMaps, scheme)
+	if err != nil {
+		return
+	}
+
+	err = c.generateSystemBackupYAMLsForServices(dir, "services", scheme)
+	if err != nil {
+		return
+	}
+
+	err = c.generateSystemBackupYAMLsForStorageClasseResources(dir, "persistentvolumes", c.ds.GetAllPersistentVolumesByStorageClasses, scheme)
```

Comment:
I just realized that this function will skip the Longhorn StorageClasses if they are not used by any existing PV. Can we blindly pick up the resources based on field `provisioner: driver.longhorn.io`?

Reply:
1. Missed backup of the storage classes. Added.
2. Not sure what you mean by blindly picking the resources? Can you elaborate?